libsolidity/analysis/ImmutableValidator.cpp

```diff
@@ -0,0 +1,252 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libsolidity/analysis/ImmutableValidator.h>
+
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace solidity::frontend;
+
+void ImmutableValidator::analyze()
+{
+	m_inConstructionContext = true;
+
+	for (VariableDeclaration const* stateVar: m_currentContract.stateVariablesIncludingInherited())
+		if (stateVar->value())
+		{
+			stateVar->value()->accept(*this);
+			solAssert(m_initializedStateVariables.insert(stateVar).second, "");
+		}
+
+	for (auto const* contract: m_currentContract.annotation().linearizedBaseContracts  | boost::adaptors::reversed)
+	{
+		m_inConstructionContext = true;
+
+		if (contract->constructor())
+		{
+			contract->constructor()->accept(*this);
+			m_visitedCallables.insert(contract->constructor());
+		}
+
+		for (std::shared_ptr<InheritanceSpecifier> const inheritSpec: contract->baseContracts())
+			if (auto args = inheritSpec->arguments())
+				ASTNode::listAccept(*args, *this);
+
+		m_inConstructionContext = false;
+
+		for (auto funcDef: contract->definedFunctions())
+			if (m_visitedCallables.find(funcDef) == m_visitedCallables.end())
+				funcDef->accept(*this);
+
+		for (auto modDef: contract->functionModifiers())
+			if (m_visitedCallables.find(modDef) == m_visitedCallables.end())
+				modDef->accept(*this);
+	}
+
+	checkAllVariablesInitialized(m_currentContract.location());
+}
+
+bool ImmutableValidator::visit(FunctionDefinition const& _functionDefinition)
+{
+	return analyseCallable(_functionDefinition);
+}
+
+
+bool ImmutableValidator::visit(ModifierDefinition const& _modifierDefinition)
+{
+	return analyseCallable(_modifierDefinition);
+}
+
+
+bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
+{
+	_memberAccess.expression().accept(*this);
+
+	if (auto funcType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type))
+		if (
+			(funcType->kind() == FunctionType::Kind::Internal ||
+			funcType->kind() == FunctionType::Kind::Declaration) &&
+			funcType->hasDeclaration()
+		)
+			if (m_visitedCallables.insert(dynamic_cast<CallableDeclaration const*>(&funcType->declaration())).second)
```

Comment:
Can you create a utility function for that? (handling "already visited and calling accept if not" and "setting visited"?)